Allow choosing which apps will get restored

grote opened the pull request:
This adds quite a bit of plumping around the restore process. E.g. we now separately store all the icons of apps we back up, as well as their names and system app status. Apps during restore are now also (mostly) sorted by name.

This then leads to a bit of changes in how we display backup status, e.g. check-boxes instead of switches, show system data instead of system apps and include launch-able system apps in list.

### Testing

* check that backup status list looks as expected
* check that apps can still be excluded from backup
* test restore process in various scenarios
  * from existing installation
  * from new install SUW
  * with system app restore and without (while restoring some shown system apps and others not)
  * check that unselected apps don't get restored
  * with APK backup enabled and without
  * check that APKs get installed fine
  * check that APKs that could not get installed or were not backed up can manually be installed
  * check that a backup made with an older version restores just as fine as one made with the current version

Closes #309, #624

grote committed:
Change app excludes from switches to checkboxes

apparently this is closer to the material design specs:
https://m2.material.io/components/checkboxes#usage

grote committed:
Allow user to choose which apps should get restored

grote committed:
Sort apps that failed to install by name

grote committed:
Show other (launchable) system apps in backup status

grote committed:
Store app icons in separate file

so they can be shown when selecting apps for restore which is before we have downloaded the APK files to extract icons from

grote committed:
Store app name and whether it is a launchable system app in metadata

this is helpful for restore, so the user can see app names when deciding which apps to restore

grote committed:
Show actual app name in app selection for restore

grote committed:
Also store icons of launchable system apps

grote committed:
Sort app selection like backup status and show sections

system data comes first and then apps

grote committed:
Cleaned up backup status list by separating system data and apps

grote committed:
Fix icon colors (again!)

see https://github.com/seedvault-app/seedvault/commit/5662409ef5c59c31557e2d9eab0a0691c5dbc00f

grote committed:
Add meta item for restoring all (internal) system apps

This way we don't need to show the long complicated list, but provide an all or nothing option at least.

grote committed:
Improve icon display when selecting apps for restore

grote committed:
Encrypt zip file with icons

While we still don't guarantee that an attacker with access to the storage can't find out which apps we use (APKs are still unencrypted after all), we go into this direction.

Also, this should make it impossible for an attacker that can modify files to replace or otherwise mess with the icons.

grote committed:
Factor out app restore selection code into new AppSelectionManager

grote committed:
Upgrade Roboeletric and bump minSdk to 34

grote committed:
Show size for (all combined) system apps when restoring

grote committed:
Add tests for AppSelectionManager

grote committed:
Re-factor and improve ApkRestore

grote committed:
don't record excluded apps in backup

grote committed:
Fix AOSP build

grote committed:
Fix backup/restore instrumentation tests

grote committed:
Fix red error color

Resources$NotFoundException: Resource ID #0x7f060258 type #0x2 is not valid

grote committed:
Don't try to install system apps without APK

during restore process. These can usually not be manually installed anyway and just clutter the list making it harder for the user to find their important apps and potential failures there.

grote committed:
Delete cached icons after restore is done

grote committed:
Factor out code into new
AppDataRestoreManager

which was in RestoreViewModel before.
Now all three steps of app restore have their own dedicated manager class making the ViewModel more readable.

grote committed:
Improve app data restore process

Apps are now restored alphabetically to be consistent with the other lists. Some irrelevant apps are hidden. Under the hood, we now use an AsyncListDiffer like in the other lists.

grote committed:
Try to fix issue where our transport doesn't get registered in emulator

grote committed:
Filter out @end@ helper package

it doesn't really exist and should probably not even be in the backup metadata. It messes up progress reporting, if we keep it in.

t-m-w commented:
> with system app restore and without (while restoring some shown system apps and others not)

When restoring, I see a "System apps" toggle under "System data", but I do not see the ability to select or deselect individual system apps. What I'm looking for is e.g. Chromium, but I don't see it. Am I missing something?

Edit: I do see some, such as Phone and Music, mixed in with other user apps (that's the part I wasn't expecting, but it's fine). It turns out that Calculator, Chromium, Clock, and some others are "Waiting to back up..." for some reason. I do have D2D on, currently.

t-m-w commented:
UX: As I scroll around the list of apps to restore, when I scroll down, including when I reach the bottom, there is no clear way to proceed with restoration, as the "Restore backup" button hides itself when you scroll down. I need to scroll up a bit to make the floating "Restore backup" button reappear, which is a bit unintuitive.

grote commented:
> When restoring, I see a "System apps" toggle under "System data", but I do not see the ability to select or deselect individual system apps. What I'm looking for is e.g. Chromium, but I don't see it. Am I missing something?
> 
> Edit: I do see some, such as Phone and Music, mixed in with other user apps (that's the part I wasn't expecting, but it's fine). It turns out that Calculator, Chromium, Clock, and some others are "Waiting to back up..." for some reason. I do have D2D on, currently.

We do show launchable system apps like regular apps now. Other non-userfacing system apps are behind that "System apps" toggle as an all or nothing.

> UX: As I scroll around the list of apps to restore, when I scroll down, including when I reach the bottom, there is no clear way to proceed with restoration, as the "Restore backup" button hides itself when you scroll down. I need to scroll up a bit to make the floating "Restore backup" button reappear, which is a bit unintuitive.

Maybe @theimpulson knows a coordinator layout behavior that works better here. Because of the long app list and the need to go through all of it, I didn't want to steal precious screen space for that button and hence let it float above the list.

t-m-w commented:
When trying to restore an older backup (from CalyxOS 5.7.1, Pixel 8a akita), I do not see app names or icons, only package names and generic icons. Waiting a while does not cause them to load.
<img src="https://github.com/seedvault-app/seedvault/assets/7275539/47688658-b7c6-4c7f-aa4a-fc43154bba56" width="240" />

grote commented:
> When trying to restore an older backup (from CalyxOS 5.7.1, Pixel 8a akita), I do not see app names or icons, only package names and generic icons. 

Yes that is expected. Only the new code does store names and icons. As we don't control old installs and old backups, there's no way we can retroactively store those things. So restoring old backups will need to live without icons and app names.

Both do exist in the APKs, if available, but downloading all APKs before the user can select something is a bit much, so we didn't do it.

t-m-w commented:
Okay, everything checks out to me except for one thing which isn't really a Seedvault problem / not related to this change. I left a bullet next to it.

* [x] check that backup status list looks as expected
* [x] check that apps can still be excluded from backup
* test restore process in various scenarios
  * [x] from existing installation
  * [x] from new install SUW
  * [x] with system app restore and without (while restoring some shown system apps and others not)
  * [x] check that unselected apps don't get restored
  * [x] with APK backup enabled and without
  * [x] check that APKs get installed fine
  * [ ] check that APKs that could not get installed or were not backed up can manually be installed
    * [CalyxOS-specific failure](https://gitlab.com/CalyxOS/calyxos/-/issues/2437): F-Droid Basic and Aurora Store are not available at this time. Seedvault launches Chromium to a Play Store page, where nothing can be done.
    * If you do an existing-install restore with `*#*#RESTORE#*#*`, it's okay, except that it immediately fails for apps which have no APKs backed up, even if those apps were installed in the interim - so you tap to install, go back without doing anything, and the app moves to the bottom with a green checkmark. Unrelated issue I think, I believe I've seen it before.
  * [x] check that a backup made with an older version restores just as fine as one made with the current version
    * System apps are not listed individually, but this is likely not possible or feasible.

grote commented:
Thanks a lot for testing @t-m-w!

> If you do an existing-install restore with *#*#RESTORE#*#*, it's okay, except that it immediately fails for apps which have no APKs backed up, even if those apps were installed in the interim - so you tap to install, go back without doing anything, and the app moves to the bottom with a green checkmark. Unrelated issue I think, I believe I've seen it before.

Right, that's unrelated and one of the items of https://github.com/seedvault-app/seedvault/issues/671

stevesoltys committed:
Skip magic package manager in assertions

We still don't know why, but after latest test fixes, the @pm@ package didn't have the the @meta@ key in the backup, but it did in restore.

stevesoltys commented on app/src/androidTest/java/com/stevesoltys/seedvault/e2e/SeedvaultLargeTest.kt:
```diff
@@ -44,6 +45,11 @@ internal abstract class SeedvaultLargeTest :
         clearTestBackups()
 
+        runCommand("bmgr enable true")
+        sleep(60_000)
```
Fine for now but this could become some sort of logcat grep... maybe possible...

grote replied:
yeah would be nice to shorten this (and other) times.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/ui/SystemData.kt:
```diff
@@ -0,0 +1,28 @@
+internal const val PACKAGE_NAME_SETTINGS = "com.android.providers.settings"
+internal const val PACKAGE_NAME_CALL_LOG = "com.android.calllogbackup"
+internal const val PACKAGE_NAME_CONTACTS = "org.calyxos.backup.contacts"
+internal const val PACKAGE_NAME_SYSTEM = "@org.calyxos.system@"
```
Is this specific to CalyxOS?

grote replied:
No, it is just a fake package name like `@pm@` and just matters at runtime, I had `@system@` first but then was afraid Android ever starts using it for something. We could use `@org.stevesoltys.seedvault.system@` as well.
